### PR TITLE
GH-3406 remove projections optimizer

### PR DESCRIPTION
**Pull request by JervenBolleman**
GitHub issue resolved: #GH-3406
Briefly describe the changes proposed in this PR:

Basic optimizer that removes projections in the algebra that don't contribute to the results.
As a first step I changed many (hopefully all) sparql parsers to always return a TupleExpr rooted in a QueryRoot.

----
PR Author Checklist (see the [contributor guidelines](https://github.com/eclipse/rdf4j/blob/main/.github/CONTRIBUTING.md) for more details):

 - [X] my pull request is [self-contained](https://rdf4j.org/documentation/developer/merge-strategy/#self-contained-changes-pull-requests-and-commits)
 - [X] I've added tests for the changes I made
 - [X] I've applied [code formatting](https://github.com/eclipse/rdf4j/blob/main/.github/CONTRIBUTING.md#code-formatting) (you can use `mvn process-resources` to format from the command line)
 - [X] I've [squashed](https://rdf4j.org/documentation/developer/squashing) my commits where necessary 
 - [X] every commit message starts with the issue number (GH-xxxx) followed by a meaningful description of the change

### CODE REVIEW

**Comment by JervenBolleman** [comment]
The getAllStatementsSparql is significantly improved

This PR
```
Benchmark                                         Mode  Cnt    Score   Error  Units
SparqlOverheadBenchmark.getAllStatementSparql     avgt    5  189.830 ± 3.560  ms/op
SparqlOverheadBenchmark.getAllStatementsDirectly  avgt    5   29.458 ± 1.421  ms/op
SparqlOverheadBenchmark.queryWithManyVars         avgt    5    0.477 ± 0.011  ms/op
```
Develop
```
Benchmark                                         Mode  Cnt    Score   Error  Units
SparqlOverheadBenchmark.getAllStatementSparql     avgt    5  257.077 ± 5.601  ms/op
SparqlOverheadBenchmark.getAllStatementsDirectly  avgt    5   26.089 ± 1.125  ms/op
SparqlOverheadBenchmark.queryWithManyVars         avgt    5    0.394 ± 0.023  ms/op
```

File: core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/ProjectionRemovealOptimizer.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
+/*******************************************************************************
+ * Copyright (c) 2019 Eclipse RDF4J contributors.
```
Comment (jeenbroekstra) [inline core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/ProjectionRemovealOptimizer.java:2]: You had this lying around? :)

File: core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/ProjectionRemovealOptimizer.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+ * </code> Does not need a projection as the inner statement pattern returns the same result.
+ * 
+ * While <code>
+ *  * SELECT ?s ?p 
+ * WHERE {?s ?p ?o }
+ * </code> Does as the statement pattern has one more variable in use than the projection.
+ * 
+ * @author Jerven Bolleman
+ */
+public class ProjectionRemovealOptimizer implements QueryOptimizer {
```
Comment (jeenbroekstra) [inline core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/ProjectionRemovealOptimizer.java:37]: Minor typo, should be `ProjectionRemovalOptimizer`

File: core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/ProjectionRemovealOptimizer.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+import org.eclipse.rdf4j.query.algebra.evaluation.QueryOptimizer;
+import org.eclipse.rdf4j.query.algebra.helpers.AbstractQueryModelVisitor;
+
+/**
+ * If a projection node in the algebra does not contribute or change the results it can be removed from the tree.
+ *
+ * For example <code>
+ * SELECT ?s ?p ?o
+ * WHERE {?s ?p ?o }
+ * </code> Does not need a projection as the inner statement pattern returns the same result.
```
Comment (jeenbroekstra) [inline core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/ProjectionRemovealOptimizer.java:28]: I haven't looked in detail but I'd like to check specifically how this works in the case of subselects - I seem to remember that we have several optimizers that look for subselects by checking if there is a (nested) projection. We'll have to make sure that those cases continue to work.

File: core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/ProjectionRemovealOptimizer.java (hunk 1 above)
Comment (JervenBolleman) [reply core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/ProjectionRemovealOptimizer.java:2]: I really need to set up my editor to add this notice for me. Will fix.

Comment (jeenbroekstra) [reply core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/ProjectionRemovealOptimizer.java:2]: Fwiw I still get it wrong regularly myself (especially on new unit tests). I wouldn't object to something in the maven formatter  that automatically checks/adds it, but I'm not sure if there is such an option.

File: core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/ProjectionRemovealOptimizer.java (hunk 3 above)
Comment (JervenBolleman) [reply core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/ProjectionRemovealOptimizer.java:28]: I think there is an other option but that would require merging #3354 . Then we can mark an algebra step as not needed and remove it from the evaluation.

File: core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/ProjectionRemovealOptimizer.java (hunk 2 above)
Comment (JervenBolleman) [reply core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/ProjectionRemovealOptimizer.java:37]: Ai. Will fix

File: core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/ProjectionRemovealOptimizer.java (hunk 3 above)
Comment (jeenbroekstra) [reply core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/ProjectionRemovealOptimizer.java:28]: Oh I'm not against this approach per se - I'd just like to be sure it doesn't break subselect-queries. My remark was from the top of my head so I may be way off base.

**Review by jeenbroekstra — APPROVED** [review]
I'm happy with this as-is. Nice work :)

**Pull request merged**